File upload accessibility changes

tombye opened the pull request:
Work to fix issues with our file upload component, spotted in a previous accessibility audit.

https://www.pivotaltracker.com/story/show/177000534

As well as this bug, I also found separate issues with some assistive technologies while testing the component. This pull request attempts to fix both.

## The current approach

It's worth explaining how the existing approach works.

<img width="378" alt="upload_current_state_diagram" src="https://user-images.githubusercontent.com/87140/173029259-6d142cf2-cae0-430d-b47a-fea44dc5bc91.png">

It involves the input[type=file] element being visually hidden and it having 3 labels:
1. the real label for the input[type=file] element. It's only visually hidden so it still applies
2. a label styled as a button which:
    - when clicked, activates the input[type=file] element
    - is styled to look focused when the input[type=file] element is focused so it looks like you can tab to it
3. a label that used to hold the name of the file selected. Now redundant since [we made the form submit when the file has been chosen](https://github.com/alphagov/notifications-admin/commit/97a3bf922580f61c81f386f9665a3f05724f4ec4)

This works for visual users with mouse/pointers and keyboard users because clicking on the button activates the file chooser:

![show_click_on_current_upload_button](https://user-images.githubusercontent.com/87140/173078667-685251dd-f413-4d3b-8d0d-33a28a3bc178.gif)

...and it 'looks like' you can tab to it:

![show_tab_to_current_upload_button](https://user-images.githubusercontent.com/87140/173079379-8659bb51-be4d-4779-8053-cd567181bd3f.gif)

The problems with it are:
- assistive technologies have problems with multiple labels
- the label styled as a button is announced as a label by screen readers, not a button
- saying 'Click upload file' to speech recognition software, to click a button with 'upload file' text, won't work because it isn't a button
- moving to the input[type=file], which makes the label button look focused, announces the input[type=file] element, which doesn't match the visuals

## Potential solutions

We could keep patching these issues individually: giving the label a 'button' role, hiding the input[type=file] element, etc... but I think it's simpler to just hide all the initial elements and add a real `<button>`. In full, this proposal is to:
1. hide all the existing elements in the upload form, visually and semantically
5. add a button
6. use JS to make clicks on the button click the input[type=file]

Effectively, all that's left in the page is a button and because it's a real `<button>` element it works as expected with all assistive technologies.

## Problems with the content you see when the file is uploading

While working on this, I noticed a few other problems with the content you see when your file is uploading.

When you click the upload button it sends your file to our S3 bucket and swaps out the upload button for another button to let you cancel this if it takes too long.

![show_full_flow_for_current_upload_button](https://user-images.githubusercontent.com/87140/173101768-aef68c70-774d-49fc-9f31-85d439273d1e.gif)

Problems with the loading state:
1. The current focus (style used by keyboard and screen reader users to see where they are) is lost because the upload button is removed
4. Screen reader users aren't told what's happening

This proposal tries to fix that by:
1. wrapping the cancel button in a paragraph with text telling you the file is uploading
2. adding some hidden text to the paragraph saying there's now a button you can use to cancel the upload
5. focusing the paragraph so focus isn't lost and the paragraph gets announced

It looks like this visually:

<img width="265" alt="upload_loading_content_visuals" src="https://user-images.githubusercontent.com/87140/172886895-c8cbd8e3-f6b4-4516-9b97-9ab142e5976f.png">

It sounds like this in screen readers:

<img width="490" alt="upload_loading_content_announcement" src="https://user-images.githubusercontent.com/87140/172886930-500884b7-f3dd-4fae-bb00-d4aec040a0a3.png">

I think this makes sense because it tells you what's happened and hints at the button being in the paragraph (called 'group' in the announcement). I plan to pair with @karlchillmaid to make it better soon.

## How do I check this?

I recommend reviewing commit-by-commit. I've checked it in all [the browsers required by the service manual](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#browsers-to-test-in-from-june-2021) and [all the assistive technologies](https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies#which-assistive-technologies-to-test-with).

The upload field used when adding a new email brand is a good page to test on.

It's also worth checking without JS and, if you feel comfortable doing it, in a screen reader. I'd also be happy to demo it with some assistive technologies if that would help.

benthorner reviewed (commented):
Even as a user who doesn't employ assistive tech, this feels better: the way the button focuses is consistent with all other buttons. 💯

I think @quis should take a look due to the design change while a file is uploading. I'll add a screenshot below to save time. One thing I noticed was that it appears to "flash" more than it did before if you upload something with an error.

<img width="374" alt="Screenshot 2022-06-21 at 17 18 25" src="https://user-images.githubusercontent.com/9029009/174850306-b9386b03-f973-4ce4-b998-9c191e9a8b76.png">

I've found two potential technical issues worth looking at before this goes in. All my other comments should be easy fixes or OK to defer / drop.

tombye commented on app/assets/javascripts/fileUpload.js:
```diff
@@ -9,13 +9,29 @@
+      this.$field = this.$form.find('.file-upload-field');
+
+      // Hide all controls except the button, semantically and visually
+      // These elements are also hidden in their CSS to override any use of the display style
```
I don't think we do actually. I started with just adding the `hidden` attribute because it is a really nice, explicit way of saying an element is semantically and visually hidden but then found:
- `display` being set in the CSS will override the visual hiding so needs to be set to `none` anyway
- it's an odd experience, visually, to have JS make something disappear whereas hiding it with the page styles (hung off the `js-enabled` class) makes it hidden from the start

...so in other words, this made sense when I was doing it but doesn't for someone reading it now so I'll remove the use of `hidden`.

tombye commented on app/assets/javascripts/fileUpload.js:
```diff
@@ -5,9 +5,26 @@
+      new window.GOVUK.Frontend.Button(this.$form[0]).init();
+
+      // move focus to the content, it is lost when the upload button is removed
+      $loadingContent.focus();
```
The yellow box is the focus state which is applied using the [:focus-visible pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible). Using :focus-visible means the focus style isn't shown if you're using a pointer (mouse/trackpad/finger on touch screens) to interact with the page. I'm guessing that that's why you're seeing what you are so be useful to have that confirmed if you can.

This is the approach we've used elsewhere as you don't tend to need to know what is focused when using pointers but do if using a keyboard or screen reader. Be interested to know if @quis feels that works with the visual design or not too, as you mentioned above.

benthorner replied:
Ah yes that seems to work consistently. I'm not sure how I managed to make it visible before as I wasn't using the keyboard to navigate; the important thing is it works when it needs to.

I'm going to resolve this one as we've got that overall design question @quis ☝️.

tombye commented on tests/javascripts/fileUpload.test.js:
```diff
@@ -53,7 +62,40 @@ describe('File upload', () => {
+    We can't simulate the user choosing a file so we test the behaviours resulting from the click
+    and onchange events.
+  */
+  test("If the 'upload' button is clicked, it should click the file upload control", () => {
```
Fair enough. I admit I wrestled with how to compose this bit as you have to do some pretty ugly things to recreate the interaction due to restrictions on how the file chooser can be activated.

tombye replied:
Now added in https://github.com/alphagov/notifications-admin/pull/4268/commits/b984358e53bd745fc88e61e3d0fe16c043a1a59f

tombye replied:
I've had a go at this in https://github.com/alphagov/notifications-admin/pull/4268/commits/1c687cd21644d953832d52826510e7872d041500

tombye commented on app/assets/stylesheets/components/file-upload.scss:
```diff
@@ -1,52 +1,34 @@
-      display: inline-block;
-      line-height: 35px;
+  // Only show a button added by JS if JS is available
+  .js-enabled &-label,
```
This is a great spot 💯. I've made some changes to show the errors which I think make sense and should replicate the existing visual design in https://github.com/alphagov/notifications-admin/pull/4268/commits/3fa33a314d53f6bf7a8e74b84fd16c168f6b439b:

<img width="382" alt="image" src="https://user-images.githubusercontent.com/87140/175059714-0fb443f6-b716-4aa3-9fe8-aae01c688780.png">

As for the point you make about the page and form-level error mismatch, this is something I'm hoping to roll into this story: https://www.pivotaltracker.com/story/show/170030139

tombye replied:
The behaviour you describe is confusing. Looking into this, it's mainly to do with our validation but some of it seems to do with Voiceover specifcally too. For anyone unfamiliar with this page, when you try and upload an invalid file you trigger a GET request for the current page which returns it, with validation added to the upload field.

In screen readers, all page loads normally start with an announcement that lists the page title and gives an overview of its contents. When we fix our validation we'll add text to the page title saying there are errors on the page, following design system guidance, so this is the first thing you hear.

Voiceover seems to not be announcing the page title and contents but, instead, jumping straight to focusing the first focusable item (the skiplink). This seems like a bug with Voiceover but this is the first time I've seen it so unsure without more research.

As for your other points, it is odd to have 2 labels in the page but I needed to associate the error text with the button and it seemed to be slightly cleaner to give it it's own label so there's a clear line between what is the hidden field and what is the new field (the button). Your testing has made one problem with that clear: that the error text is **replacing** the button's accessible name (normally its button text), not adding to it.

tombye replied:
I've had a go at fixing the button's label by adding the existing accessible name to the start of it in https://github.com/alphagov/notifications-admin/pull/4268/commits/e58e33d41a44dd09dc47d9b691fac1c55aa7f628. You now hear this:

<img width="639" alt="file-upload-new-accessible-name" src="https://user-images.githubusercontent.com/87140/175296889-ef6f819a-308f-4b97-b2c0-80dc06c5ce33.png">

benthorner replied:
That's better 🏅. Seems to be consistent with [the design system](https://design-system.service.gov.uk/components/file-upload/) too, except that normal buttons don't have error states - just a file upload picker button.

I think I understand how the two labels work as well: the no-JS one never gets read out unless we're running without JS, whereas the label injected by JS will get read out (including visually hidden text).

The screen reader issue sounds potentially bad. Do you think it's worth a low impact / high effort bug card to investigate further?

tombye replied:
Ben has left but it's worth addressing the last point here. I think myself and Ben were both experiencing Voiceover not announcing the normal overview of the page when it loads. I noticed my version of Voiceover has the 'Speak web summary' option unticked:

<img width="458" alt="voiceover_utility_web_settings" src="https://user-images.githubusercontent.com/87140/181287293-a588617d-dd10-4fd4-9b12-0c4bf5f56e42.png">

Asking around I heard most people share this. Turning it on also doesn't seem to make much of a difference, in Chrome or Safari, as Voiceover starts talking through the page before the summary is complete, or even started in some cases. I'm going to continue investigating but the thing to note here is that it isn't associated with these changes so is out of scope for this PR.

tombye commented:
Thanks for the excellent review @benthorner, I hope I've addressed all the comments made. The one outstanding thing here seems to be the content shown while your file is uploading. These changes include my go at how to solve the problem we had in a way that makes sense but, so as to make it easier to redesign this, the requirements are:
- the focus the upload button has when you activate it must go somewhere, it currently disappears when the button does
- it must be clear to users of screen readers who don't interact with the visual interface what's happening when you click the upload button and that they can use the cancel button to cancel the upload

tombye commented:
I've reverted out all but the following changes, to move work out that doesn't have an obvious solution:
- reduce labels to 1
- make the upload button use a `<button>` tag

Full details of the things removed, why they were removed and what still needs doing are in the [previous commit](https://github.com/alphagov/notifications-admin/pull/4268/commits/2719997037dafec9c2bd8363153d6673621b99f3).

I'm going to write the remaining work up into stories and link to them here.

This should be considered ready for checking again.

tombye commented:
After some consideration, I think this is too complicated for anyone new to review. I'm closing this in favour of https://github.com/alphagov/notifications-admin/pull/4310 which is a rewrite of the final changes into something with all the suggested corrections but without all the context here.
